custom_components/bhyve/binary_sensor.py

```diff
@@ -0,0 +1,80 @@
+"""Support for Orbit BHyve sensors."""
+import logging
+
+from . import BHyveDeviceEntity
+from .const import (
+    DATA_BHYVE,
+    DEVICE_FLOOD,
+    EVENT_FS_ALARM,
+    EVENT_DEVICE_IDLE,
+)
+from homeassistant.components.binary_sensor import (
+    DEVICE_CLASS_MOISTURE,
+    BinarySensorEntity,
+)
+from .pybhyve.errors import BHyveError
+from .util import orbit_time_to_local_time
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_platform(hass, config, async_add_entities, _discovery_info=None):
+    """Set up BHyve sensors based on a config entry."""
+    bhyve = hass.data[DATA_BHYVE]
+
+    sensors = []
+    devices = await bhyve.devices
+    for device in devices:
+        if device.get("type") == DEVICE_FLOOD:
```

Comment:
@disforw can you please provide a dump of the `device` json for the Flood Sensor? Thank you

Reply:
Do you mean an API dump?

Reply:
Yes please, I'd like to see the contents of the JSON payload

Reply:
@disforw any chance you can give me a sample here? Thanks :)

Reply:
@disforw I'm keen to get this over the line, but I can't make progress without a copy of the API dump. Can you please provide me with this and I'll get it done?

Reply:
Hey, sorry ive been MIA. Tell me what you need and I will provide, but I dont know how to get it, so point me in the right direction

Reply:
Thanks mate.

Step 1. Add the following to `configuration.yaml` and then restart Home Assistant.
```
logger:
  logs:
    custom_components.bhyve: debug
```

Step 2: Check Home Assistant logs

1. Open `home-assistant.log`
2. Identify lines which begin with:
```log
DEBUG (MainThread) [custom_components.bhyve] Devices:
DEBUG (MainThread) [custom_components.bhyve] Programs:
```

Copy these two lines and paste them below.

```log
DEBUG (MainThread) [custom_components.bhyve] Devices:
DEBUG (MainThread) [custom_components.bhyve] Programs:
```

Reply:
2021-11-07 18:49:46 DEBUG (MainThread) [custom_components.bhyve] Devices: [{"local_ui_settings": {"lockout_enabled": false, "lights_enabled": true}, "last_connected_at": "2021-11-07T20:46:31.165Z", "address": "REDACTED", "water_sense_mode": "auto", "timezone": {"dst_offset": 3600, "raw_offset": -18000, "timezone_id": "America/New_York", "timezone_name": "Eastern Daylight Time"}, "full_location": "REDACTED", "weather_forecast_location_id": "5a3026074f0c1bde5df55526", "firmware_version": "0039", "name": "Bhyve Controller", "type": "sprinkler_timer", "battery": null, "restricted_frequency": {"type": "odd", "days": [0, 1, 2, 3, 4, 5, 6], "restricted_times": []}, "weather_delay_thresholds": {"precip_prob": 30, "precip_in": 0.125, "wind_speed_mph": 20, "freeze_temp_f": 37}, "updated_at": "2021-11-07T20:46:31.165Z", "reference": "4467550320f0", "mac_address": "4467550320f0", "weather_station_id": "5a3026074f0c1bde5df55527", "mesh_id": "5b0749eb4f0c43d0e0a57e55", "status": {"run_mode": "auto", "next_start_programs": [], "rain_sensors": [], "rain_delay_overridden_at": "2019-06-03T01:53:48.372Z", "watering_status": null, "rain_delay_suggested_at": "2018-09-17T09:56:06.103Z", "rain_delay": 0, "flow_sensor": null, "next_start_time": "1970-01-01T00:00:00-05:00", "rain_delay_started_at": "2021-11-07T01:44:22.000Z"}, "wifi_version": 0, "id": "5b0749ea4f0c43d0e0a57e53", "num_stations": 6, "zones": [{"station": 1, "slope_grade": 0, "sun_shade": "partial_shade", "name": "Front East Side", "landscape_type": "warm_season_turf", "num_sprinklers": 2, "soil_type": "loam", "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "sprinkler_type": "rotor", "flow_data": {}}, {"station": 2, "slope_grade": 0, "sun_shade": "partial_shade", "name": "Front West Side", "landscape_type": "warm_season_turf", "num_sprinklers": 2, "soil_type": "sandy_loam", "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "sprinkler_type": "rotor", "flow_data": {}}, {"station": 3, "slope_grade": 0, "sun_shade": "full_sun", "name": "Front Sidewalk", "landscape_type": "warm_season_turf", "num_sprinklers": 2, "soil_type": "sandy_loam", "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "sprinkler_type": "rotor", "flow_data": {}}, {"station": 4, "slope_grade": 0, "sun_shade": "partial_shade", "name": "Backyard", "landscape_type": "warm_season_turf", "num_sprinklers": 2, "soil_type": "sandy_loam", "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "sprinkler_type": "rotor", "flow_data": {}}], "user_id": "5b0749c04f0c43d0e0a57e46", "google_home_status": {"will_report_state": true, "user_id": "5b0749c04f0c43d0e0a57e46", "request_id": null}, "device_gateway_topic": "devices-6", "hardware_version": "WT25G2-0001", "is_connected": true, "location": "REDACTED", "created_at": "2018-05-24T23:25:30.149Z", "suggested_start_time": "06:00", "et_id": "5644ff05ad27a7777ff6994c"}, {"last_connected_at": "2021-11-07T20:46:29.494Z", "address": "REDACTED", "timezone": {"dst_offset": 3600, "raw_offset": -18000, "timezone_id": "America/New_York", "timezone_name": "Eastern Daylight Time"}, "full_location": "REDACTED", "firmware_version": "0027", "name": "Bhyve WiFi Hub", "type": "bridge", "updated_at": "2021-11-07T20:46:29.494Z", "reference": "44675524b51a", "mac_address": "44675524b51a", "wifi_version": 0, "id": "61280a794f0c6c9c9fad8ca1", "num_stations": 0, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "BH1G2-0001", "is_connected": true, "location": "REDACTED", "created_at": "2021-08-26T21:41:13.925Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3"}, {"last_connected_at": "2021-08-26T21:49:07.126Z", "location_name": "Laundry Room", "firmware_version": "14", "temp_alarm_thresholds": {"high_temp_f": 85, "low_temp_f": 55}, "name": "Washing Machine", "type": "flood_sensor", "battery": {"percent": 100}, "updated_at": "2021-11-07T22:58:08.739Z", "reference": "44675560cba2", "mac_address": "44675560cba2", "status": {"last_flood_alarm_at": "2021-08-27T18:24:54.717Z", "rssi": -74, "temp_f": 68.0, "provisioned": true, "phy": "le_1m_1000", "temp_alarm_status": "ok", "status_updated_at": "2021-11-07T22:58:08.738Z", "flood_alarm_status": "ok"}, "id": "61280bdf4f0cad87f43fe230", "auto_shutoff": false, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "fs1-0001", "is_connected": true, "created_at": "2021-08-26T21:47:11.293Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3", "address": "REDACTED", "full_location": "REDACTED", "location": "REDACTED"}, {"last_connected_at": "2021-09-17T07:24:50.989Z", "location_name": "Furnace Room", "firmware_version": "14", "temp_alarm_thresholds": {"high_temp_f": 90, "low_temp_f": 60}, "name": "Boiler Room", "type": "flood_sensor", "battery": {"percent": 100}, "updated_at": "2021-11-07T23:36:59.719Z", "reference": "44675560cc50", "mac_address": "44675560cc50", "status": {"last_flood_alarm_at": "2021-10-17T14:58:47.710Z", "rssi": -71, "temp_f": 69.8, "provisioned": true, "phy": "le_1m_1000", "temp_alarm_status": "ok", "status_updated_at": "2021-11-07T23:36:59.718Z", "flood_alarm_status": "ok"}, "id": "612ad9134f0c6c9c9faddbba", "auto_shutoff": false, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "fs1-0001", "is_connected": true, "created_at": "2021-08-29T00:47:15.061Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3", "address": "REDACTED", "full_location": "REDACTED", "location": "REDACTED"}, {"last_connected_at": "2021-10-27T14:36:22.334Z", "location_name": "Bathroom", "firmware_version": "14", "temp_alarm_thresholds": {"high_temp_f": 85, "low_temp_f": 55}, "name": "Kids Bathroom", "type": "flood_sensor", "battery": {"percent": 100}, "updated_at": "2021-11-07T23:36:38.997Z", "reference": "44675560a354", "mac_address": "44675560a354", "status": {"rssi": -85, "temp_f": 68.0, "provisioned": true, "phy": "le_1m_1000", "temp_alarm_status": "ok", "status_updated_at": "2021-11-07T23:36:38.997Z", "flood_alarm_status": "ok"}, "id": "612bc70d4f0c6c9c9fadee8b", "auto_shutoff": false, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "fs1-0001", "is_connected": true, "created_at": "2021-08-29T17:42:37.019Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3", "address": "REDACTED", "full_location": "REDACTED", "location": "REDACTED"}, {"last_connected_at": "2021-08-29T18:11:29.253Z", "location_name": "Crawlspace", "firmware_version": "14", "temp_alarm_thresholds": {"high_temp_f": 90, "low_temp_f": 50}, "name": "Crawlspace Hatch", "type": "flood_sensor", "battery": {"percent": 100}, "updated_at": "2021-11-07T23:46:59.546Z", "reference": "44675560c9f8", "mac_address": "44675560c9f8", "status": {"last_flood_alarm_at": "2021-09-01T02:53:36.354Z", "rssi": -74, "temp_f": 66.2, "provisioned": true, "phy": "le_1m_1000", "temp_alarm_status": "ok", "status_updated_at": "2021-11-07T23:46:59.545Z", "flood_alarm_status": "ok"}, "id": "612bcd864f0c6c9c9fadeffb", "auto_shutoff": false, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "fs1-0001", "is_connected": true, "created_at": "2021-08-29T18:10:14.218Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3", "address": "REDACTED", "full_location": "REDACTED", "location": "REDACTED"}, {"last_connected_at": "2021-09-20T04:33:18.988Z", "location_name": "Kitchen", "firmware_version": "14", "temp_alarm_thresholds": {"high_temp_f": 100, "low_temp_f": -5}, "name": "Kitchen Freezer", "type": "flood_sensor", "battery": {"percent": 98}, "updated_at": "2021-11-07T23:19:31.225Z", "reference": "44675560c9b1", "mac_address": "44675560c9b1", "status": {"rssi": -73, "temp_f": 6.800000000000001, "provisioned": true, "phy": "le_1m_1000", "temp_alarm_status": "ok", "status_updated_at": "2021-11-07T23:19:31.225Z", "flood_alarm_status": "ok", "last_temp_alarm_at": "2021-09-05T22:24:10.917Z"}, "id": "612c2d244f0cad87f440596b", "auto_shutoff": false, "user_id": "5b0749c04f0c43d0e0a57e46", "device_gateway_topic": "devices-2", "hardware_version": "fs1-0001", "is_connected": true, "created_at": "2021-08-30T00:58:12.080Z", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3", "address": "REDACTED", "full_location": "REDACTED", "location": "REDACTED"}, {"last_connected_at": "2021-09-03T23:26:20.568Z", "address": "REDACTED", "water_sense_mode": "off", "timezone": {"dst_offset": 3600, "raw_offset": -18000, "timezone_id": "America/New_York", "timezone_name": "Eastern Daylight Time"}, "full_location": "REDACTED", "weather_forecast_location_id": "5a3026074f0c1bde5df55526", "firmware_version": "58", "name": "Bhyve Hose", "type": "sprinkler_timer", "manual_preset_runtime_sec": 600, "battery": {"percent": 100, "mv": 3170}, "restricted_frequency": null, "weather_delay_thresholds": {"precip_prob": 30, "precip_in": 0.125, "wind_speed_mph": 20, "freeze_temp_f": 37}, "updated_at": "2021-11-07T20:46:29.505Z", "reference": "446755900576", "mac_address": "446755900576", "weather_station_id": "5a3026074f0c1bde5df55527", "status": {"run_mode": "off", "rssi": -57, "rain_sensors": [], "provisioned": true, "phy": "le_1m_1000", "rain_delay": 0, "status_updated_at": "2021-09-07T00:54:21.711Z", "flow_sensor": null, "rain_delay_started_at": "2021-09-03T13:03:48.000Z"}, "id": "613157ad4f0cf51668e7bc87", "num_stations": 2, "zones": [{"station": 1, "name": "Hose Left", "num_sprinklers": 0, "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "flow_data": {}}, {"station": 2, "name": "Hose Right", "num_sprinklers": 0, "catch_cup_volumes": [], "catch_cup_run_time": 0, "smart_watering_enabled": false, "flow_data": {}}], "user_id": "5b0749c04f0c43d0e0a57e46", "google_home_status": null, "device_gateway_topic": "devices-2", "hardware_version": "HT32-0001", "is_connected": true, "location": "REDACTED", "created_at": "2021-09-02T23:01:01.129Z", "suggested_start_time": null, "et_id": "5644ff05ad27a7777ff6994c", "network_topology_id": "61280a7a4f0c6c9c9fad8ca3"}]
2021-11-07 18:49:46 DEBUG (MainThread) [custom_components.bhyve] Programs: [{"name": "Unnamed Program 1", "frequency": {"type": "odd"}, "updated_at": "2021-05-25T19:59:32.902Z", "start_times": ["05:30"], "id": "608946454f0c5f2af2a8a616", "budget": 80, "is_smart_program": false, "device_id": "5b0749ea4f0c43d0e0a57e53", "program": null, "run_times": [{"run_time": 15.0, "station": 1}, {"run_time": 15.0, "station": 2}, {"run_time": 10.0, "station": 3}, {"run_time": 15.0, "station": 4}], "enabled": false, "created_at": "2021-04-28T11:25:57.011Z"}, {"update_via": "wifi", "name": "Primary", "frequency": {"type": "odd"}, "is_basic_program": false, "updated_at": "2021-07-29T15:17:05.058Z", "updated_via": "wifi", "start_times": ["05:30"], "id": "6089464e4f0c161d5a10a419", "budget": 80, "is_smart_program": false, "device_id": "5b0749ea4f0c43d0e0a57e53", "program": "a", "run_times": [{"run_time": 15.0, "station": 1}, {"run_time": 15.0, "station": 2}, {"run_time": 10.0, "station": 3}, {"run_time": 15.0, "station": 4}], "enabled": false, "created_at": "2021-04-28T11:26:06.416Z"}, {"update_via": "wifi", "name": "Backyard", "frequency": {"type": "days", "days": [0, 1, 2, 3, 4, 5, 6]}, "is_basic_program": false, "updated_at": "2021-05-25T19:59:32.906Z", "updated_via": "wifi", "start_times": ["06:00", "08:00"], "id": "608947094f0c9e0599a73234", "budget": 100, "is_smart_program": false, "device_id": "5b0749ea4f0c43d0e0a57e53", "program": "b", "run_times": [{"run_time": 10.0, "station": 1}, {"run_time": 10.0, "station": 2}, {"run_time": 10.0, "station": 3}], "enabled": false, "created_at": "2021-04-28T11:29:13.870Z"}, {"update_via": "wifi", "name": "Front Yard", "frequency": {"type": "days", "days": [0, 1, 2, 3, 4, 5, 6]}, "is_basic_program": false, "updated_at": "2021-10-01T11:44:29.150Z", "updated_via": "wifi", "start_times": ["06:00"], "id": "60ac16274f0cc5e78a30c4e4", "budget": 100, "is_smart_program": false, "device_id": "5b0749ea4f0c43d0e0a57e53", "program": "c", "run_times": [{"run_time": 15.0, "station": 1}, {"run_time": 15.0, "station": 2}, {"run_time": 15.0, "station": 3}], "enabled": false, "created_at": "2021-05-24T21:09:59.347Z"}]
2021-11-07 18:49:46 INFO (MainThread) [homeassistant.setup] Setup of domain bhyve took 0.8 seconds

Reply:
Just so you know what you’re looking at, I have a main controller, a 2 zone hose valve, the usb Wi-Fi controller, and 4 flood sensors.